ocaml-lsp-server/src/inference.ml

```diff
@@ -66,25 +107,193 @@ let open_document_from_file (state : State.t) uri =
         in
         Some doc)
 
-let infer_intf (state : State.t) doc =
-  match Document.kind doc with
+let infer_intf (state : State.t) intf_doc =
+  match Document.kind intf_doc with
   | `Other ->
     Code_error.raise "the provided document is not a merlin source." []
   | `Merlin m when Document.Merlin.kind m = Impl ->
     Code_error.raise "the provided document is not an interface." []
   | `Merlin m ->
     Fiber.of_thunk (fun () ->
-        let intf_uri = Document.uri doc in
+        let intf_uri = Document.uri intf_doc in
         let impl_uri =
-          Document.get_impl_intf_counterparts (Some m) intf_uri |> List.hd
+          Document.get_impl_intf_counterparts (Some m) intf_uri |> List.hd_exn
         in
-        let* impl =
+        let* impl_opt =
           match Document_store.get_opt state.store impl_uri with
           | Some impl -> Fiber.return (Some impl)
           | None -> open_document_from_file state impl_uri
         in
-        match impl with
+        match impl_opt with
         | None -> Fiber.return None
-        | Some impl ->
-          let+ res = infer_intf_for_impl impl in
+        | Some impl_doc ->
+          let+ res = infer_missing_intf_for_impl impl_doc intf_doc in
           Some res)
+
+(** Extracts an [Ident.t] from all variants that have one at the top level. For
+    many of the other variants, it would be possible to extract a list of IDs,
+    but that's not needed for the update-signatures code action. *)
+let top_level_id (item : Typedtree.signature_item) =
+  match item.sig_desc with
+  | Typedtree.Tsig_value { val_id; _ } -> Some val_id
+  | Typedtree.Tsig_module { md_id; _ } -> md_id
+  | Typedtree.Tsig_modsubst { ms_id; _ } -> Some ms_id
+  | Typedtree.Tsig_modtype { mtd_id; _ } -> Some mtd_id
+  | Typedtree.Tsig_modtypesubst { mtd_id; _ } -> Some mtd_id
+  | Typedtree.Tsig_type _
+  | Typedtree.Tsig_typesubst _
+  | Typedtree.Tsig_typext _
+  | Typedtree.Tsig_exception _
+  | Typedtree.Tsig_recmodule _
+  | Typedtree.Tsig_open _
+  | Typedtree.Tsig_include _
+  | Typedtree.Tsig_class _
+  | Typedtree.Tsig_class_type _
+  | Typedtree.Tsig_attribute _ -> None
```

Comment:
If we want to relax `ocaml-lsp`'s dependency on the `Typedtree` on the long term we should really put such utilities in `merlin-lib` instead. 

Also, I am curious about the reason why classes' ids are not needed for the update-signature action ?